xls/integrator/ir_integrator.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+#include "xls/integrator/ir_integrator.h"
+
+#include "xls/ir/ir_parser.h"
+
+namespace xls {
+
+// Return the name of 'function', prepended with the 'package' name.
+std::string GetParsableQualifiedFunctionName(Package* package,
+                                             const Function* function) {
+  return "PKGzzz" + package->name() + "zzzFNzzz" + function->name();
+}
+
+absl::Status IntegrationBuilder::CopySourcesToIntegrationPackage() {
+  std::string integration_package_str = "package IntegrationPackage\n";
+
+  // Dump IR for original functions with qualified names.
+  for (const auto* source : original_package_source_functions_) {
+    // Copy functions to temporary package to avoid modifying
+    // original functions.
+    std::string srctmp_package_str = "package srctmp\n";
+    srctmp_package_str.append(source->DumpIr(/*recursive=*/true));
+    XLS_ASSIGN_OR_RETURN(auto srctmp_package,
+                         Parser::ParsePackage(srctmp_package_str));
+
+    // Change all function names before dumping any function so that
+    // all function invocations refer to the new function name.
+    for (const auto& function : srctmp_package->functions()) {
+      function->SetName(
+          GetParsableQualifiedFunctionName(source->package(), function.get()));
+    }
+
+    // Dump functions to common IR string.
+    for (const auto& function : srctmp_package->functions()) {
```

Comment:
I'm guessing dumping functions from different packages into the same IR file will fail because of duplicate node ids which must be unique in the package scope. Using Function::Clone avoids this issue.

Reply:
Done.